controller/run.go

```diff
@@ -6,147 +6,71 @@ package controller
 
 import (
 	"context"
+	"errors"
 	"fmt"
-	"net"
 	"os"
 	"os/signal"
-	"sync"
 	"syscall"
 
-	"google.golang.org/grpc"
-	"google.golang.org/grpc/credentials"
-
 	"github.com/elastic/elastic-agent-client/v7/pkg/client"
 	"github.com/elastic/elastic-agent-libs/logp"
 	"github.com/elastic/elastic-agent-shipper/config"
-	"github.com/elastic/elastic-agent-shipper/monitoring"
-	"github.com/elastic/elastic-agent-shipper/output"
-	"github.com/elastic/elastic-agent-shipper/queue"
-	"github.com/elastic/elastic-agent-shipper/server"
-
-	pb "github.com/elastic/elastic-agent-shipper-client/pkg/proto"
 )
 
 // LoadAndRun loads the config object and runs the gRPC server
 func LoadAndRun() error {
-	agentClient, _, err := client.NewV2FromReader(os.Stdin, client.VersionInfo{Name: "elastic-agent-shipper", Version: "v2"})
-	if err != nil {
-		return fmt.Errorf("error reading control config from agent: %w", err)
+	cfg, err := config.ReadConfigFromFile()
+	switch {
+	case err == nil:
+		return RunUnmanaged(cfg)
+	case errors.Is(err, config.ErrConfigIsNotSet):
+		return RunManaged(cfg)
+	default:
+		return err
 	}
-	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
-	err = runController(ctx, agentClient)
-
-	return err
 }
 
-// handle shutdown of the shipper
-func handleShutdown(stopFunc func(), done doneChan) {
-	var callback sync.Once
+// RunUnmanaged runs the shipper out of a local config file without using the control protocol.
+func RunUnmanaged(cfg config.ShipperConfig) error {
 	log := logp.L()
-	// On termination signals, gracefully stop the Beat
-	sigc := make(chan os.Signal, 1)
-	signal.Notify(sigc, syscall.SIGINT, syscall.SIGTERM, syscall.SIGHUP)
-	go func() {
-		for {
-			select {
-			case <-done:
-				log.Debugf("Shutting down from agent controller")
-				callback.Do(stopFunc)
-				return
-			case sig := <-sigc:
-				switch sig {
-				case syscall.SIGINT, syscall.SIGTERM:
-					log.Debug("Received sigterm/sigint, stopping")
-				case syscall.SIGHUP:
-					log.Debug("Received sighup, stopping")
-				}
-				callback.Do(stopFunc)
-				return
-			}
-		}
-
-	}()
-}
-
-// Run starts the gRPC server
-func (c *clientHandler) Run(cfg config.ShipperConfig, unit *client.Unit) error {
-	log := logp.L()
-
-	// When there is queue-specific configuration in ShipperConfig, it should
-	// be passed in here.
-	queue, err := queue.New(cfg.Queue)
-	if err != nil {
-		return fmt.Errorf("couldn't create queue: %w", err)
-	}
-
-	// Make a placeholder console output to read the queue's events
-	out := output.NewConsole(queue)
-	out.Start()
-
-	lis, err := net.Listen("tcp", fmt.Sprintf("localhost:%d", cfg.Server.Port))
-	if err != nil {
-		return fmt.Errorf("failed to listen: %w", err)
-	}
-
-	monHandler, err := loadMonitoring(cfg, queue)
+	runner, err := NewServerRunner(cfg)
 	if err != nil {
-		return fmt.Errorf("error loading outputs: %w", err)
+		return err
 	}
+	done := make(doneChan)
+	go func() {
+		err = runner.Start()
```

Comment:
Looks like we return this err below, any possibility we get an odd situation where a signal might happen before Start finishes?

Reply:
Unless I missed something, this should be okay because of the `startMutex` that would not allow to execute `Close` until `Start` finished.